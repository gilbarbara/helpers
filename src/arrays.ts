import { PlainObject } from '@gilbarbara/types';

import { SortFunction } from './types';

/**
 * Create a sequential array of numbers
 */
export function createArray(size: number, start = 1) {
  return Array.from({ length: size }, (_, index) => index + start);
}

/**
 * Get a random item from an array
 */
export function getRandomItem<T>(input: T[]) {
  return input[Math.floor(Math.random() * input.length)];
}

/**
 * Sort an array of numbers using a quick sort algorithm
 */
export function quickSort<T extends string | number>(input: T[], comparator = sortComparator): T[] {
  // Create a sortable array to return.
  const output = [...input];

  // Recursively sort sub-arrays.
  const recursiveSort = (start: number, end: number) => {
    // If this sub-array is empty, it's sorted.
    if (end - start < 1) {
      return;
    }

    const pivotValue = output[end];
    let splitIndex = start;

    for (let index = start; index < end; index++) {
      const sort = comparator(output[index], pivotValue);

      // This value is less than the pivot value.
      if (sort === -1) {
        // If the element just to the right of the split index,
        //   isn't this element, swap them.
        if (splitIndex !== index) {
          const temp = output[splitIndex];

          output[splitIndex] = output[index];
          output[index] = temp;
        }

        // Move the split index to the right by one,
        //   denoting an increase in the less-than sub-array size.
        splitIndex++;
      }

      // Leave values that are greater than or equal to
      //   the pivot value where they are.
    }

    // Move the pivot value to between the split.
    output[end] = output[splitIndex];
    output[splitIndex] = pivotValue;

    // Recursively sort the less-than and greater-than arrays.
    recursiveSort(start, splitIndex - 1);
    recursiveSort(splitIndex + 1, end);
  };

  // Sort the entire array.
  recursiveSort(0, input.length - 1);

  return output;
}

/**
 * Remove duplicates from the array
 */
export function removeDuplicates<T = unknown>(input: T[]): T[] {
  return [...new Set(input)];
}

/**
 * Shuffle an array using the Fisher-Yates algorithm
 */
export function shuffle<T = unknown>(input: T[]) {
  let currentIndex = input.length;
  let randomIndex;
  let temporaryValue;
  const output = [...input];

  while (currentIndex) {
    randomIndex = Math.floor(Math.random() * currentIndex--);
    temporaryValue = output[currentIndex];
    output[currentIndex] = output[randomIndex];
    output[randomIndex] = temporaryValue;
  }

  return output;
}

/**
 * Sort an array with localeCompare
 */
export function sortByLocaleCompare(
  key?: string,
  options: Intl.CollatorOptions & { descending?: boolean } = {},
): SortFunction {
  const { descending, ...compareOptions } = options;

  if (key) {
    if (descending) {
      return <T extends PlainObject<any>>(left: PlainObject<T>, right: PlainObject<T>) =>
        right[key].toLowerCase().localeCompare(left[key].toLowerCase(), undefined, compareOptions);
    }

    return <T extends PlainObject<any>>(left: PlainObject<T>, right: PlainObject<T>) =>
      left[key].toLowerCase().localeCompare(right[key].toLowerCase(), undefined, compareOptions);
  }

  if (descending) {
    return <T extends string>(left: T, right: T) =>
      right.toLowerCase().localeCompare(left.toLowerCase(), undefined, compareOptions);
  }

  return <T extends string>(left: T, right: T) =>
    left.toLowerCase().localeCompare(right.toLowerCase(), undefined, compareOptions);
}

/**
 * Sort an array by primitive values
 */
export function sortByPrimitive<T extends number | boolean>(
  key?: string,
  descending = false,
): SortFunction {
  const firstComparator = descending ? 1 : -1;
  const secondComparator = descending ? -1 : 1;

  if (key) {
    return <P extends PlainObject<any>>(left: PlainObject<P>, right: PlainObject<P>) => {
      if (left[key] === right[key]) {
        return 0;
      }

      return left[key] < right[key] ? firstComparator : secondComparator;
    };
  }

  return <P extends T>(left: P, right: P) => {
    if (left === right) {
      return 0;
    }

    return left < right ? firstComparator : secondComparator;
  };
}

/**
 * Basic sort comparator
 */
export function sortComparator(left: string | number, right: string | number) {
  if (left < right) {
    return -1;
  }

  if (left > right) {
    return 1;
  }

  return 0;
}

export function splitIntoChunks<T>(input: T[], chunkSize: number = 25): T[][] {
  if (!Array.isArray(input)) {
    throw new TypeError('expected an array for the first argument');
  }

  const chunks: T[][] = [];

  for (let index = 0; index < Math.ceil(input.length / chunkSize); index++) {
    chunks.push(input.slice(index * chunkSize, (index + 1) * chunkSize));
  }

  return chunks;
}
