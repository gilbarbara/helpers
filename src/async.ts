import 'isomorphic-fetch';

import is from 'is-lite';

import { CorsOptions, PollOptions, RequestError, RequestOptions } from './types';

/**
 * Format a CORS response
 */
export function cors(data: any, statusCodeOrOptions: number | CorsOptions = 200) {
  const {
    allowCredentials = true,
    allowedHeaders = [],
    methods = ['GET'],
    origin = '*',
    responseHeaders = undefined,
    statusCode = 200,
  } = is.number(statusCodeOrOptions) ? { statusCode: statusCodeOrOptions } : statusCodeOrOptions;

  const allowMethods = [...methods, 'OPTIONS'];
  const allowHeaders = [
    ...new Set([
      'Accept-Version',
      'Accept',
      'Authorization',
      'Content-Length',
      'Content-MD5',
      'Content-Type',
      'Date',
      'x-amz-date',
      'x-amz-security-token',
      'X-Api-Version',
      'X-CSRF-Token',
      'X-Requested-With',
      ...allowedHeaders,
    ]),
  ];
  let exposedHeaders;

  if (responseHeaders) {
    exposedHeaders = {
      'Access-Control-Expose-Headers': Object.keys(responseHeaders).join(','),
      ...responseHeaders,
    };
  }

  return {
    body: JSON.stringify(data),
    headers: {
      'Access-Control-Allow-Origin': origin,
      'Access-Control-Allow-Credentials': allowCredentials,
      'Access-Control-Allow-Methods': allowMethods.join(','),
      'Access-Control-Allow-Headers': allowHeaders.join(','),
      ...exposedHeaders,
    },
    statusCode,
  };
}

/**
 *
 * @param condition
 * @param options
 */
export async function poll(condition: () => boolean, options: PollOptions = {}): Promise<void> {
  const { delay = 1, maxRetries = 5 } = options;
  let retries = 0;

  while (!condition() && retries <= maxRetries) {
    // eslint-disable-next-line no-await-in-loop
    await sleep(delay);
    retries++;
  }

  if (retries >= maxRetries) {
    throw new Error('Timeout');
  }
}

/**
 * Make async requests
 */
export async function request<D = any>(url: string, options: RequestOptions = {}): Promise<D> {
  const { body, headers, method = 'GET' }: RequestOptions = options;

  if (!url) {
    throw new Error('URL is required');
  }

  const params: RequestOptions = {
    headers: {
      Accept: 'application/json',
      'Content-Type': 'application/json',
      ...headers,
    },
    method,
  };

  if (body) {
    params.body = JSON.stringify(body);
  }

  return fetch(url, params).then(async response => {
    const text = await response.text();
    let content: any;

    try {
      content = JSON.parse(text);
    } catch {
      content = text;
    }

    if (response.status > 299) {
      const error = new Error(response.statusText) as RequestError;

      error.status = response.status;
      error.response = content;

      throw error;
    } else {
      return content;
    }
  });
}

/**
 * Block execution
 */
export function sleep(seconds = 1) {
  return new Promise(resolve => {
    setTimeout(resolve, seconds * 1000);
  });
}
