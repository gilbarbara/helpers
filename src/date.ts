import is from 'is-lite';

import { pluralize } from './strings';
import { TimeSinceOptions } from './types';

export const MINUTE = 60;
export const HOUR = MINUTE * 60;
export const DAY = HOUR * 24;
export const WEEK = DAY * 7;
export const MONTH = DAY * 30;
export const YEAR = 365 * DAY;

export function isIsoDate(input: string) {
  if (!/\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}.\d{3}Z/.test(input)) {
    return false;
  }

  const date = new Date(input);

  return date.toISOString() === input;
}

/**
 * Check if the input is a valid date.
 */
export function isoDate(input?: string | number) {
  if (typeof input !== 'undefined') {
    return new Date(input).toISOString();
  }

  return new Date().toISOString();
}

/**
 * Check if the input is a valid date.
 */
export function isValidDate(input: string | number | Date) {
  const date = is.date(input) ? input : new Date(input);

  return !Number.isNaN(date.getTime());
}

/**
 * Returns the unixtime (in seconds).
 */
export function now() {
  return Math.floor(Date.now() / 1000);
}

/**
 * Returns how much time has passed since the input.
 */
export function timeSince(input: Date | string | number, options?: TimeSinceOptions) {
  const {
    day = 'day',
    days,
    hour = 'hour',
    hours,
    minute = 'minute',
    minutes,
    month = 'month',
    months,
    prefix,
    second = 'second',
    seconds,
    skipWeeks = false,
    suffix = 'ago',
    week = 'week',
    weeks,
    year = 'year',
    years,
  } = options || {};
  const date = typeof input === 'number' ? input : timestamp(input);
  const diff = now() - date;
  let quantity = Math.floor(diff / YEAR);

  const output = (value: number, caption: string) => {
    let result = `${value} ${caption}`;

    if (prefix) {
      result = `${prefix} ${result}`;
    }

    if (suffix) {
      result = `${result} ${suffix}`;
    }

    return result;
  };

  if (quantity >= 1) {
    return output(quantity, pluralize(quantity, year, years));
  }

  quantity = Math.floor(diff / MONTH);

  if (quantity > 1) {
    return output(quantity, pluralize(quantity, month, months));
  }

  if (!skipWeeks) {
    quantity = Math.floor(diff / WEEK);

    if (quantity > 1) {
      return output(quantity, pluralize(quantity, week, weeks));
    }
  }

  quantity = Math.floor(diff / DAY);

  if (quantity >= 1) {
    return output(quantity, pluralize(quantity, day, days));
  }

  quantity = Math.floor(diff / HOUR);

  if (quantity >= 1) {
    return output(quantity, pluralize(quantity, hour, hours));
  }

  quantity = Math.floor(diff / MINUTE);

  if (quantity > 1) {
    return output(quantity, pluralize(quantity, minute, minutes));
  }

  return output(Math.floor(diff), pluralize(diff, second, seconds));
}

/**
 * Get the timestamp (in seconds) for a date.
 */
export function timestamp(input?: Date | string) {
  if (!input) {
    return now();
  }

  const date = typeof input === 'string' ? new Date(input) : input;

  return Math.floor(date.getTime() / 1000);
}
