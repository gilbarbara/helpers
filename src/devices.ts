/**
 * Detect if the device is in dark mode
 */
export function isDarkMode() {
  return !!window.matchMedia && window.matchMedia('(prefers-color-scheme: dark)').matches;
}

/**
 * Detect if the device supports touch events
 */
export function isTouchDevice() {
  return !!window.matchMedia && window.matchMedia('(pointer: coarse)').matches;
}

/**
 * Detect if the user prefers reduced motion
 */
export function prefersReducedMotion() {
  return !!window.matchMedia && window.matchMedia('prefers-reduced-motion').matches;
}
