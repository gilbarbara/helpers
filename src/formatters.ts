import { cleanupNumericString } from './strings';
import { FormatDateLocaleOptions, FormatMoneyOptions } from './types';

/**
 * Format boolean into a Yes/No string
 */
export function formatBoolean(input: boolean) {
  return input ? 'Yes' : 'No';
}

/**
 * Format string into a CPF
 */
export function formatCPF(value: string) {
  const clearValue = cleanupNumericString(value);

  if (clearValue.length >= 10) {
    return `${clearValue.slice(0, 3)}.${clearValue.slice(3, 6)}.${clearValue.slice(
      6,
      9,
    )}-${clearValue.slice(9, 11)}`;
  } else if (clearValue.length >= 7) {
    return `${clearValue.slice(0, 3)}.${clearValue.slice(3, 6)}.${clearValue.slice(6, 9)}`;
  } else if (clearValue.length >= 4) {
    return `${clearValue.slice(0, 3)}.${clearValue.slice(3, 6)}`;
  }

  return clearValue;
}

/**
 * Format date ISO string using locale
 */
export function formatDateLocale(input: string, options: FormatDateLocaleOptions = {}) {
  const { locale = 'en-GB', showTime = false } = options;

  const formatOptions: Intl.DateTimeFormatOptions = {
    year: '2-digit',
    month: '2-digit',
    day: '2-digit',
  };

  if (showTime) {
    formatOptions.hour = '2-digit';
    formatOptions.minute = '2-digit';
  }

  return new Date(input).toLocaleDateString(locale, formatOptions);
}

/**
 * Format number into money string
 */
export function formatMoney(input: number, options: FormatMoneyOptions = {}) {
  const { decimalChar = '.', showCents = false, symbol = '$', thousandsChar = ',' } = options;

  const isNegative = input < 0;
  const value = Math.abs(input);
  const [amount, cents] = value.toFixed(2).split('.');
  const padStart = amount.length > 3 ? amount.length % 3 : 0;
  const initial = amount.slice(0, padStart);
  const remain = amount.slice(padStart).replace(/(\d{3})(?=\d)/g, `$1${thousandsChar}`);

  let formatted = `${remain}`;

  if (initial) {
    formatted = `${initial}${thousandsChar}${remain}`;
  }

  if (cents !== '00' || showCents) {
    formatted += `${decimalChar}${cents}`;
  }

  return `${isNegative ? '-' : ''}${symbol}${formatted}`;
}

/**
 * Format string into a brazilian phone
 */
export function formatPhoneBR(input: string) {
  const phone = input.replace(/\D/g, '');

  if (phone.length === 8) {
    return phone.replace(/^(\d{4})(\d{4}).*/, '$1-$2');
  }

  if (phone.length === 9) {
    return phone.replace(/^(\d{5})(\d{4}).*/, '$1-$2');
  }

  if (phone.length === 10 || phone.length === 11) {
    return phone.replace(/^(\d{2})(\d{4,5})(\d{4}).*/, '($1) $2-$3');
  }

  return phone;
}

/**
 * Format string into a US phone
 */
export function formatPhoneUS(input: string) {
  const phone = input.replace(/\D/g, '');

  if (phone.length === 10) {
    return phone.replace(/^(\d{3})(\d{3})(\d{4}).*/, '($1) $2-$3');
  }

  if (phone.length === 11 && phone.startsWith('1')) {
    return phone.replace(/^1(\d{3})(\d{3})(\d{4}).*/, '+1 ($1) $2-$3');
  }

  return phone;
}

/**
 * Format string into a zip code
 */
export function formatPostalCodeBR(value: string) {
  const clearValue = cleanupNumericString(value);

  if (clearValue.length >= 6) {
    return `${clearValue.slice(0, 5)}-${clearValue.slice(5, 8)}`;
  }

  return clearValue;
}
