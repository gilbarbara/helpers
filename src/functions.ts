/**
 * Decouple methods from objects
 */
// eslint-disable-next-line @typescript-eslint/ban-types
export function demethodize(fn: Function) {
  return (parameter: any, ...rest: any[]) => fn.apply(parameter, rest);
}

/**
 * Measure function execution time
 */
// eslint-disable-next-line @typescript-eslint/ban-types
export async function measureExecutionTime<T = any>(callback: Function): Promise<T> {
  const start = performance.now();

  const result = await callback();

  const end = performance.now();
  const total = end - start;

  // eslint-disable-next-line no-console
  console.log(`Completed in ${Math.ceil(total)} milliseconds`);

  return result as T;
}

/**
 * A function that does nothing.
 */
export function noop() {
  return undefined;
}

/**
 * Creates a function that will only be called once.
 * Repeat calls return the value of the first invocation.
 */
export function once<T extends (...arguments_: Array<any>) => any>(fn: T): T {
  let done = false;
  let result: ReturnType<T>;

  return function Fn(...arguments_: Parameters<T>): ReturnType<T> {
    if (!done) {
      done = true;
      result = fn(...arguments_);
    }

    return result as ReturnType<T>;
  } as T;
}

/**
 * Combine multiple functions into one.
 * The output of each function is passed as the input to the next.
 */
export function pipe<T>(...fns: Array<(argument: T) => T>) {
  return (input: T) => fns.reduce((previousValue, fn) => fn(previousValue), input);
}
