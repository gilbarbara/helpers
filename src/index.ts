export * from './arrays';
export * from './async';
export * from './date';
export * from './misc';
export * from './numbers';
export * from './objects';
export * from './strings';

export * from './types';
