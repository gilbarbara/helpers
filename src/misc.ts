import { pad } from './numbers';
import { LoggerOptions, UniqueOptions } from './types';

/**
 * Copy a string to the clipboard
 */
export async function copyToClipboard(input: string) {
  try {
    await navigator.clipboard.writeText(input);
  } catch {
    return false;
  }

  return true;
}

/**
 * Decouple methods from objects
 */
// eslint-disable-next-line @typescript-eslint/ban-types
export function demethodize(fn: Function) {
  return (parameter: any, ...rest: any[]) => fn.apply(parameter, rest);
}

/**
 * Check if a string is a valid JSON
 */
export function isJSON(input: string) {
  try {
    JSON.parse(input);
  } catch {
    return false;
  }

  return true;
}

/**
 * Throw an error if the parameter isn't provided
 */
export function isRequired(input = 'parameter', Constructable = TypeError) {
  throw new Constructable(`"${input}" is required`);
}

/**
 * Log grouped messages to the console
 */
export function logger(type: string, title: string, data: any, options: LoggerOptions = {}) {
  /* eslint-disable no-console */
  const { collapsed = true, hideTimestamp = false, skip = false, typeColor = 'gray' } = options;
  const groupMethod = collapsed ? console.groupCollapsed : console.group;
  const date = new Date();
  const parts = [`%c ${type}`];
  const styles = [`color: ${typeColor}; font-weight: lighter;`, 'color: inherit;'];

  if (!hideTimestamp) {
    styles.push('color: gray; font-weight: lighter;');
  }

  const time = `${pad(date.getHours(), 2)}:${pad(date.getMinutes(), 2)}:${pad(
    date.getSeconds(),
    2,
  )}`;

  parts.push(`%c${title}`);

  if (!hideTimestamp) {
    parts.push(`%c@ ${time}`);
  }

  /* istanbul ignore else */
  if (!skip) {
    groupMethod(parts.join(' '), ...styles);
    console.log(data);
    console.groupEnd();
  }
  /* eslint-enable */
}

/**
 * Measure function execution time
 */
// eslint-disable-next-line @typescript-eslint/ban-types
export async function measureExecutionTime<T = any>(callback: Function): Promise<T> {
  const start = performance.now();

  const result = await callback();

  const end = performance.now();
  const total = end - start;

  // eslint-disable-next-line no-console
  console.log(`Completed in ${Math.ceil(total)} milliseconds`);

  return result as T;
}

/**
 * A function that does nothing.
 */
export function noop() {
  return undefined;
}

/**
 * Returns the value or null
 */
export function nullify<T>(value: T) {
  return value ?? null;
}

export function popupCenter(url: string, title: string, width: number, height: number) {
  const { screen } = window;
  let { screenLeft } = window;
  let { screenTop } = window;
  let screenWidth = window.innerWidth;
  let screenHeight = window.innerHeight;

  /* istanbul ignore next */
  if (typeof screenLeft === 'undefined') {
    // @ts-ignore
    screenLeft = screen.left;
  }

  /* istanbul ignore next */
  if (typeof screenTop === 'undefined') {
    // @ts-ignore
    screenTop = screen.top;
  }

  /* istanbul ignore next */
  if (!screenWidth) {
    screenWidth = document.documentElement.clientWidth
      ? document.documentElement.clientWidth
      : screen.width;
  }

  /* istanbul ignore next */
  if (!screenHeight) {
    screenHeight = document.documentElement.clientHeight
      ? document.documentElement.clientHeight
      : screen.height;
  }

  const left = screenWidth / 2 - width / 2 + screenLeft;
  const top = screenHeight / 2 - height / 2 + screenTop;

  const popup = window.open(
    url,
    title,
    `menubar=no,location=no,resizable=no,scrollbars=yees,status=no,width=${width},height=${height},top=${top}, left=${left}`,
  );

  /* istanbul ignore else */
  if (popup) {
    popup.focus();
  }

  return popup;
}

/**
 * Return a unique string
 */
export function unique(length = 8, options: UniqueOptions = {}): string {
  const {
    includeLowercase = true,
    includeNumbers = true,
    includeSymbols = false,
    includeUppercase = true,
  } = options;
  const lowercase = 'abcdefghijklmnopqrstuvwxyz';
  const uppercase = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ';
  const numbers = '0123456789';
  const symbols = '!?@#$%^&*+_-=:.~';
  // const similarCharacters = /[ilLI|`oO0]/g;

  let characters = '';

  if (includeLowercase) {
    characters += lowercase;
  }

  if (includeUppercase) {
    characters += uppercase;
  }

  if (includeNumbers) {
    characters += numbers;
  }

  if (includeSymbols) {
    characters += symbols;
  }

  let result = '';

  for (let index = length; index > 0; --index) {
    result += characters[Math.round(Math.random() * (characters.length - 1))];
  }

  return result;
}

/**
 * Returns an UUID v4 string.
 */
export function uuid() {
  /* eslint-disable no-bitwise */
  return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, c => {
    const r = (Math.random() * 16) | 0;
    const v = c === 'x' ? r : (r & 0x3) | 0x8;

    return v.toString(16);
  });
  /* eslint-enable */
}
