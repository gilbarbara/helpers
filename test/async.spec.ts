import { cors, poll, request, sleep } from '../src';

describe('cors', () => {
  it.each([
    {
      name: 'data',
      data: { name: 'John Doe' },
    },
    {
      name: 'methods, origin and allowedHeaders',
      data: null,
      options: {
        allowedHeaders: ['x-test'],
        methods: ['POST' as const],
        origin: 'https://example.com',
        statusCode: 201,
      },
    },
    {
      name: 'responseHeaders',
      data: [{ name: 'John Doe' }],
      options: {
        responseHeaders: {
          Link: '</?page=10>; rel="current", </?page=10>; rel="last", </?page=9>; rel="prev"',
          'X-Items-Total': '938',
          'X-Items-Per-Page': '100',
          'X-Items-From': '901',
          'X-Items-To': '938',
        },
      },
    },
  ])('with "$name" should return properly', ({ data, options }) => {
    expect(cors(data, options)).toMatchSnapshot();
  });
});

describe('poll', () => {
  it('should resolve before the timeout', () => {
    let isReady = false;

    setTimeout(() => {
      isReady = true;
    }, 300);

    return poll(() => isReady).then(d => {
      expect(d).toBeUndefined();
    });
  });

  it('should reject after max retries', () => {
    let isReady = false;

    setTimeout(() => {
      isReady = true;
    }, 500);

    return poll(() => isReady, { delay: 0.1 }).catch(error => {
      // eslint-disable-next-line jest/no-conditional-expect
      expect(error.message).toBe('Timeout');
    });
  });
});

describe('request', () => {
  it('should throw without "url"', async () => {
    // @ts-ignore
    await expect(request()).rejects.toThrow('URL is required');
  });

  it('should handle errors', async () => {
    // @ts-ignore
    fetch.mockResponseOnce(() =>
      Promise.resolve({
        status: 404,
        body: 'Not Found',
        statusText: 'Something failed',
      }),
    );

    await expect(request('https://jsonplaceholder.typicode.com/posts/1234567890')).rejects.toThrow(
      'Something failed',
    );
  });

  it('should handle success', async () => {
    // @ts-ignore
    fetch.mockResponseOnce(
      JSON.stringify({
        body: 'quia et suscipit recusandae consequuntur expedita et cum reprehenderit molestiae ut ut quas totam nostrum rerum est autem sunt rem eveniet architecto',
        id: 1,
        title: 'Sunt aut facere repellat provident occaecati excepturi optio reprehenderit',
        userId: 1,
      }),
    );
    expect(await request('https://jsonplaceholder.typicode.com/posts/1')).toMatchSnapshot();
  });

  it('should handle options', async () => {
    // @ts-ignore
    fetch.mockResponseOnce(
      JSON.stringify({
        title: 'foo',
        body: 'bar',
        id: 101,
        userId: 1,
      }),
    );

    expect(
      await request('https://jsonplaceholder.typicode.com/posts', {
        method: 'POST',
        body: {
          title: 'foo',
          body: 'bar',
          userId: 1,
        },
      }),
    ).toMatchSnapshot();
  });
});

describe('sleep', () => {
  const setTimeoutSpy = jest.spyOn(global, 'setTimeout');

  afterAll(() => {
    jest.restoreAllMocks();
  });

  it('should halt the execution', async () => {
    // eslint-disable-next-line unicorn/consistent-function-scoping
    const fn = async () => {
      await sleep(0.1);

      return 'finished';
    };

    expect(await fn()).toBe('finished');
    expect(setTimeoutSpy).toHaveBeenCalledWith(expect.any(Function), 100);
  });
});
