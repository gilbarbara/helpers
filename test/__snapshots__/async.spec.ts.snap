// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cors with "data" should return properly 1`] = `
Object {
  "body": "{\\"name\\":\\"John Doe\\"}",
  "headers": Object {
    "Access-Control-Allow-Credentials": true,
    "Access-Control-Allow-Headers": "Accept-Version,Accept,Authorization,Content-Length,Content-MD5,Content-Type,Date,x-amz-date,x-amz-security-token,X-Api-Version,X-CSRF-Token,X-Requested-With",
    "Access-Control-Allow-Methods": "GET,OPTIONS",
    "Access-Control-Allow-Origin": "*",
  },
  "statusCode": 200,
}
`;

exports[`cors with "methods, origin and allowedHeaders" should return properly 1`] = `
Object {
  "body": "null",
  "headers": Object {
    "Access-Control-Allow-Credentials": true,
    "Access-Control-Allow-Headers": "Accept-Version,Accept,Authorization,Content-Length,Content-MD5,Content-Type,Date,x-amz-date,x-amz-security-token,X-Api-Version,X-CSRF-Token,X-Requested-With,x-test",
    "Access-Control-Allow-Methods": "POST,OPTIONS",
    "Access-Control-Allow-Origin": "https://example.com",
  },
  "statusCode": 201,
}
`;

exports[`cors with "responseHeaders" should return properly 1`] = `
Object {
  "body": "[{\\"name\\":\\"John Doe\\"}]",
  "headers": Object {
    "Access-Control-Allow-Credentials": true,
    "Access-Control-Allow-Headers": "Accept-Version,Accept,Authorization,Content-Length,Content-MD5,Content-Type,Date,x-amz-date,x-amz-security-token,X-Api-Version,X-CSRF-Token,X-Requested-With",
    "Access-Control-Allow-Methods": "GET,OPTIONS",
    "Access-Control-Allow-Origin": "*",
    "Access-Control-Expose-Headers": "Link,X-Items-Total,X-Items-Per-Page,X-Items-From,X-Items-To",
    "Link": "</?page=10>; rel=\\"current\\", </?page=10>; rel=\\"last\\", </?page=9>; rel=\\"prev\\"",
    "X-Items-From": "901",
    "X-Items-Per-Page": "100",
    "X-Items-To": "938",
    "X-Items-Total": "938",
  },
  "statusCode": 200,
}
`;

exports[`request should handle options 1`] = `
Object {
  "body": "bar",
  "id": 101,
  "title": "foo",
  "userId": 1,
}
`;

exports[`request should handle success 1`] = `
Object {
  "body": "quia et suscipit recusandae consequuntur expedita et cum reprehenderit molestiae ut ut quas totam nostrum rerum est autem sunt rem eveniet architecto",
  "id": 1,
  "title": "Sunt aut facere repellat provident occaecati excepturi optio reprehenderit",
  "userId": 1,
}
`;
